File: daily/Binary Tree/543-Diameter-of-binary-tree.go
package daily

var maxDiameter int

func diameterOfBinaryTree(root *TreeNode) int {
	maxDiameter = 0
	dfsDiameterOfBinaryTree(root)
	return maxDiameter
}

func dfsDiameterOfBinaryTree(node *TreeNode) int {
	if node == nil {
		return 0
	}

	left := dfsDiameterOfBinaryTree(node.Left)
	right := dfsDiameterOfBinaryTree(node.Right)
	maxDiameter = max(maxDiameter, left+right)
	return 1 + max(left, right)
}

File: daily/Binary Tree/543-Diameter-of-binary-tree.py
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxDiameter = 0

        def dfs(node):
            if not node:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)

            self.maxDiameter = max(self.maxDiameter, left + right)
            return 1 + max(left, right)
        dfs(root)
        return self.maxDiameter